lib/matplotlib/cbook/__init__.py

```diff
@@ -2002,6 +2002,47 @@ def _array_perimeter(arr):
     ))
 
 
+def _unfold(arr, axis, size, step):
+    """Append an extra dimension containing sliding windows along ``axis``.
+
+    All windows are of size ``size`` and begin with every ``step`` elements.
+
+    Parameters
+    ----------
+    arr : ndarray, shape (N_1, ..., N_k)
+        The input array
+
+    Returns
+    -------
+    windows : ndarray, shape (N_1, ..., 1 + (N_axis-size)/step, ..., N_k, size)
+    """
+    new_shape = list(arr.shape) + [size]
+    new_strides = list(arr.strides) + [arr.strides[axis]]
+    new_shape[axis] = (new_shape[axis] - size) // step + 1
+    new_strides[axis] = new_strides[axis] * step
+    return np.lib.stride_tricks.as_strided(arr,
+                                           shape=new_shape,
+                                           strides=new_strides,
+                                           writeable=False)
+
+
+def _array_patch_perimeters(x, rstride, cstride):
+    """Extract perimeters of patches from ``arr``.
+
+    Extracted patches are of size ``(rstride + 1) x (cstride + 1)`` and share
+    perimeters with their neighbors.
+    """
+    assert rstride > 0 and cstride > 0
+    assert (x.shape[0] - 1) % rstride == 0
+    assert (x.shape[1] - 1) % cstride == 0
+    upper = _unfold(x[:-1:rstride, :-1], 1, cstride, cstride)
```

Comment:
would something like the following work?
```
up_and_low = _unfold(x[::rstride], 1, cstride, cstride)
up = up_and_low[:-1]
low = up_and_low[1:]
...
concatenate([up, right, low[..., ::-1], left[::-1]])
```
(haven't checked the indices)
then you can make both left and right be open-intervals (up and low are both closed)
(whereas here you have all of them half-open, which *think* is slightly harder to follow?)

Reply:
Nah, either way is fine, I just never get reversed strides right on the first time.  May be worth leaving a comment explaining the four half-open intervals though (took me a while to understand it).